src/cephadm/cephadm

```diff
@@ -4363,24 +4363,64 @@ def command_bootstrap(ctx):
 
-        with open(ctx.apply_spec) as f:
+        # funcs to process spec file
+        def _parse_yaml_docs(f: Iterable[str]) -> Iterator[List[str]]:
+            current_doc = []  # type: List[str]
             for line in f:
-                if 'hostname:' in line:
-                    line = line.replace('\n', '')
-                    split = line.split(': ')
-                    if split[1] != hostname:
-                        logger.info('Adding ssh key to %s' % split[1])
-
-                        ssh_key = '/etc/ceph/ceph.pub'
-                        if ctx.ssh_public_key:
-                            ssh_key = ctx.ssh_public_key.name
-                        out, err, code = call_throws(ctx, ['sudo', '-u', ctx.ssh_user, 'ssh-copy-id', '-f', '-i', ssh_key, '-o StrictHostKeyChecking=no', '%s@%s' % (ctx.ssh_user, split[1])])
+                if '---' in line:
+                    if current_doc:
+                        yield current_doc
+                    current_doc = []
+                else:
+                    current_doc.append(line.rstrip())
+            if current_doc:
+                yield current_doc
+
+        def _parse_yaml_obj(doc: List[str]) -> Dict[str, str]:
+            # note: this only parses the first layer of yaml
+            obj = {}  # type: Dict[str, str]
+            current_key = ''
+            for line in doc:
+                if line.startswith(' '):
+                    obj[current_key] += line.strip()
+                elif line.endswith(':'):
+                    current_key = line.strip(':')
+                    obj[current_key] = ''
+                else:
+                    current_key, val = line.split(':')
+                    obj[current_key] = val.strip()
+            return obj
+
+        def parse_yaml_objs(f: Iterable[str]) -> Iterator[Dict[str, str]]:
+            for d in _parse_yaml_docs(f):
+                yield _parse_yaml_obj(d)
+
+        # copy ssh key to host specs from yaml
+        ssh_key = '/etc/ceph/ceph.pub'
+        if ctx.ssh_public_key:
+            ssh_key = ctx.ssh_public_key.name
+        with open('spec.yml') as f:
+            for spec in parse_yaml_objs(f):
+                if spec.get('service_type') == 'host':
+                    if hostname != spec['hostname']:
+                        if 'addr' in spec:
+                            addr = spec['addr']
+                        else:
+                            addr = spec['hostname']
+                        out, err, code = call(ctx, ['sudo', '-u', ctx.ssh_user, 'ssh-copy-id', '-f', '-i', ssh_key, '-o StrictHostKeyChecking=no', '%s@%s' % (ctx.ssh_user, addr)])
+                        if code:
+                            logger.info('\nCopying ssh key to host %s at address %s failed!\n' % (spec['hostname'], addr))
+                        else:
+                            logger.info('Added ssh key to host %s at address %s\n' % (spec['hostname'], addr))
```

Comment:
Could we maybe put this in its own little function right above like we do for the yaml parsing? Something like _distribute_ssh_keys that takes a host spec and the bootstrap hostname and copies the keys like this does? It's not really a big deal but it would be nice to avoid a chunk of five level deep indentation code when we can.